Snakefile

```diff
@@ -1,11 +1,42 @@
+import copy
 from os import environ
 from socket import getfqdn
 from getpass import getuser
+from snakemake.logging import logger
 from snakemake.utils import validate
+import time
+
+# Store the user's configuration prior to loading defaults, so we can check for
+# reused subsampling scheme names in the user's config. We need to make a deep
+# copy because Snakemake will deep merge the subsampling dictionary later,
+# modifying the values of a reference or shallow copy.
+user_subsampling = copy.deepcopy(config.get("subsampling", {}))
```

Comment:
is the config even defined at this point?  I thought the default parameters are loaded here, and the user's config is loaded ~L49-57.

Reply:
Yeah, `config` is defined as an empty dictionary initially. If the user provides a `--configfile` command line argument or a `configfile` attribute in their profile config, the files referenced in that argument get loaded into the `config` dictionary.

Strangely, though, these files don't appear to get immediately deep merged if there are multiple arguments. For example, the "getting started" profile (`my_profiles/getting_started/config.yaml`) specifies the default parameters and its own build parameters. The keys from both files appear in the `config` dictionary, but the values from the custom build parameters appear to override the defaults instead of getting merged. For example, the default parameters specify these in the `files` config attribute:

```yaml
files:
  include: "defaults/include.txt"
  exclude: "defaults/exclude.txt"
  reference: "defaults/reference_seq.gb"
  outgroup: "defaults/outgroup.fasta" # is this arg still used? file doesn't exist in gh repo
  ordering: "defaults/color_ordering.tsv"
  color_schemes: "defaults/color_schemes.tsv"
  auspice_config: "defaults/auspice_config.json"
  lat_longs: "defaults/lat_longs.tsv"
  description: "defaults/description.md"
  clades: "defaults/clades.tsv"
  legacy_clades: "nextstrain_profiles/nextstrain/legacy_clades.tsv"
  subclades: "defaults/subclades.tsv"
```

And the getting started profile also defines a `files` attribute with:

```yaml
files:
  auspice_config: "my_profiles/example/my_auspice_config.json"
  description: "my_profiles/example/my_description.md"
```

But when I inspect the `config` dictionary first thing in the `Snakefile`, only the "getting started" `files` entries are in there. When I inspect the `config` dictionary after the `configfile` declaration in the `Snakefile`, all of the config attributes have been deep merged.

So, I guess the downside to this new approach is that it depends on some magical behavior of Snakemake that may not persist in the future.

Reply:
Ohh, I was completely mistaken.  The later loading is for snakemake rules.

My reading of [the snakemake docs](https://snakemake.readthedocs.io/en/v4.5.0/snakefiles/configuration.html#standard-configuration) is that if `--configfile` is specified, then the defaults are not loaded.  So I guess the only case there is the deep merge is if someone specifies `--config`?

Reply:
I think [the deep merge happens for all combinations of config inputs](https://snakemake.readthedocs.io/en/stable/snakefiles/configuration.html#standard-configuration) except when someone provides a `--config` that conflicts with an existing `configfile` entry, the `--config` value will overwrite the other values.

> Note that any values parsed into the config dictionary with any of above mechanisms are merged, i.e., all keys defined via a configfile statement, or the –configfile and –config command line arguments will end up in the final config dictionary, but if two methods define the same key, command line overwrites the configfile statement.

But it seems like the deep merge doesn't happen until after the `configfile` directive has been processed in the `Snakefile`. I can see how this could be a headache from the Snakemake developers' perspective as much as it from ours.

Do you know if nextflow behaves similarly?